src/mc-fle2-rfds.c

```diff
@@ -0,0 +1,629 @@
+         break;
+      case MATCH_EXPRESSION:
+         if (!parse_match_expression (doc, str, &op, status)) {
+            goto fail;
+         }
+         break;
+      case UNKNOWN:
+      default:
+         CLIENT_ERR ("unexpected unknown expression type");
+         goto fail;
+      }
+
+      switch (op.op_type) {
+      case FLE2RangeOperator_kGt:
+         if (spec->lower.set) {
+            CLIENT_ERR (
+               "unexpected duplicate bound %s: %s", op.op_type_str, str);
+            goto fail;
+         }
+         spec->lower.set = true;
+         spec->lower.value = op.value;
+         break;
+      case FLE2RangeOperator_kGte:
+         if (spec->lower.set) {
+            CLIENT_ERR (
+               "unexpected duplicate bound %s: %s", op.op_type_str, str);
+            goto fail;
+         }
+         spec->lower.set = true;
+         spec->lower.value = op.value;
+         spec->lower.included = true;
+         break;
+      case FLE2RangeOperator_kLt:
+         if (spec->upper.set) {
+            CLIENT_ERR (
+               "unexpected duplicate bound %s: %s", op.op_type_str, str);
+            goto fail;
+         }
+         spec->upper.set = true;
+         spec->upper.value = op.value;
+         break;
+      case FLE2RangeOperator_kLte:
+         if (spec->upper.set) {
+            CLIENT_ERR (
+               "unexpected duplicate bound %s: %s", op.op_type_str, str);
+            goto fail;
+         }
+         spec->upper.set = true;
+         spec->upper.value = op.value;
+         spec->upper.included = true;
+         break;
+      case FLE2RangeOperator_kNone:
+      default:
+         CLIENT_ERR ("unsupported operator type %s: %s", op.op_type_str, str);
+         goto fail;
+         break;
+      }
+
+      if (spec->field) {
+         if (0 != strcmp (spec->field, op.field)) {
+            CLIENT_ERR ("unexpected field mismatch. Expected all fields to be "
+                        "%s, but got %s",
+                        spec->field,
+                        op.field);
+            goto fail;
+         }
+      }
+
+      spec->nOps++;
+
+      if (spec->nOps == 1) {
+         spec->firstOp = op.op_type;
+      } else if (spec->nOps == 2) {
+         spec->secondOp = op.op_type;
+      } else {
+         CLIENT_ERR (
+            "expected 1 or 2 operators, got > 2: %s: %s", op.op_type_str, str);
+         goto fail;
+      }
+
+      spec->field = op.field;
+   }
+
+   ok = true;
+fail:
+   bson_free (str);
+   return ok;
+}
+
+// mc_makeRangeFindPlaceholder creates a placeholder to be consumed by
+// libmongocrypt for encryption.
+bool
+mc_makeRangeFindPlaceholder (mc_makeRangeFindPlaceholder_args_t args,
+                             _mongocrypt_buffer_t *out,
+                             mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (out);
+   BSON_ASSERT (status || true);
+
+   bool ok = false;
+   bson_t *edgesInfo = bson_new ();
+   bson_t *v = bson_new ();
+   bson_t *p = bson_new ();
```

Comment:
Would it be worth it to just copy in the `bson-dsl.h` file? It should "just work" with the older libbson that is used by libmongocrypt, and may simplify the building/parsing aspects of this file.

Reply:
I think the DSL is likely to simplify this logic. Though given the timeline, I prefer doing as a separate PR to save time. The remaining work for Range Support is unlikely to require much parsing / building.